### PR TITLE
Dynamic set options

### PR DESCRIPTION
**Pull request by shuzhang1989**
as title.
it may be useful for us to tune / debug db issues in live.

### CODE REVIEW

File: rocksdb_admin/admin_handler.h [hunk 1]
```diff
@@ -88,7 +94,8 @@ class AdminHandler : virtual public AdminSvIf {
   std::unique_ptr<ApplicationDBManager> db_manager_;
   RocksDBOptionsGeneratorType rocksdb_options_;
   // Lock to synchronize async_tm_backupDB(), async_tm_restoreDB(),
-  // async_tm_closeDB() and async_tm_changeDBRoleAndUpStream().
+  // async_tm_closeDB(), async_tm_changeDBRoleAndUpStream() and
+  // async_tm_setDBOptions()
   common::ObjectLock<std::string> db_admin_lock_;
 };
```
Comment (newpoo) [inline rocksdb_admin/admin_handler.h]: Can you also add async_tm_clearDB() and async_tm_addS3SstFilesToDB() ?

Or just replace it with something like "Lock to synchronize DB admin operations at per DB granularity"

File: rocksdb_admin/admin_handler.cpp [hunk 2]
```diff
@@ -647,4 +647,36 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
...
+      SetDBOptionsResponse>>> callback,
+    std::unique_ptr<SetDBOptionsRequest> request) {
+  std::unordered_map<string, string> options;
+  for (const auto& option_pair : request->options) {
+    options.emplace(option_pair);
+  }
+  for (const auto& db_name : request->db_names) {
+    db_admin_lock_.Lock(db_name);
+    SCOPE_EXIT { db_admin_lock_.Unlock(db_name); };
+    auto adb = getDB(db_name, nullptr);
```
Comment (newpoo) [inline rocksdb_admin/admin_handler.cpp]: Can we rename adb to db to make it consistent with other code in the file.

File: rocksdb_admin/admin_handler.cpp [hunk 3]
```diff
@@ -647,4 +647,36 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
...
+    if (adb == nullptr) {
+      LOG(ERROR) << "Failed to get db: " << db_name;
+      continue;
+    }
+    auto db = adb -> rocksdb();
+    if (db == nullptr) {
+      LOG(ERROR) << "Failed to get db: " << db_name;
+      continue;
+    }
+    // Assume we always use default column family
```
Comment (newpoo) [inline rocksdb_admin/admin_handler.cpp]: we don't need to check if rocksdb() == nullptr. If that's the case, we'd prefer crashing it.
we can simply use auto status = db->rocksdb()->SetOptions(options).

File: rocksdb_admin/admin_handler.cpp [hunk 4]
```diff
@@ -647,4 +647,36 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
   callback->result(AddS3SstFilesToDBResponse());
 }
 
+void AdminHandler::async_tm_setDBOptions(
+    std::unique_ptr<apache::thrift::HandlerCallback<std::unique_ptr<
+      SetDBOptionsResponse>>> callback,
+    std::unique_ptr<SetDBOptionsRequest> request) {
+  std::unordered_map<string, string> options;
+  for (const auto& option_pair : request->options) {
+    options.emplace(option_pair);
```
Comment (newpoo) [inline rocksdb_admin/admin_handler.cpp]: nit:
 auto& option_pair
and
emplace(std::move(options_pair))

File: rocksdb_admin/admin_handler.cpp [hunk 5]
```diff
@@ -647,4 +647,36 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
...
+  for (const auto& db_name : request->db_names) {
+    db_admin_lock_.Lock(db_name);
+    SCOPE_EXIT { db_admin_lock_.Unlock(db_name); };
+    auto adb = getDB(db_name, nullptr);
+    if (adb == nullptr) {
+      LOG(ERROR) << "Failed to get db: " << db_name;
+      continue;
+    }
+    auto db = adb -> rocksdb();
+    if (db == nullptr) {
```
Comment (newpoo) [inline rocksdb_admin/admin_handler.cpp]: for this case, we need to send back an exception. We can fill the exception in getDB().

File: rocksdb_admin/rocksdb_admin.thrift [hunk 6]
```diff
@@ -109,6 +109,40 @@ struct AddS3SstFilesToDBResponse {
...
+  #  "target_file_size_base",
+  #  "target_file_size_multiplier",
+  #  "max_bytes_for_level_base",
+  #  "max_bytes_for_level_multiplier",
+  #  "verify_checksums_in_compaction(false),
+  #  "max_sequential_skip_in_iterations",
+  #  "paranoid_file_checks(false),
+  #  "report_bg_io_stats(false),
+  #  "compression"
+  #  "min_partial_merge_operands"
```
Comment (shuzhang1989) [inline rocksdb_admin/rocksdb_admin.thrift]: Sure, it makes more sense,

File: rocksdb_admin/rocksdb_admin.thrift [hunk 7]
```diff
@@ -163,4 +174,10 @@ ClearDBResponse clearDB(1:ClearDBRequest request)
 AddS3SstFilesToDBResponse addS3SstFilesToDB(1:AddS3SstFilesToDBRequest request)
   throws (1:AdminException e)
 
+/*
+ * Set mutable DB options.
+ * The option map in request will be passed down to Rocksdb::DB::SetOptions().
+ */
+SetDBOptionsResponse setDBOptions(SetDBOptionsRequest request)
```
Comment (newpoo) [inline rocksdb_admin/rocksdb_admin.thrift]: 1:SetDBOptionsRequest ?

**Pull request merged**